crates/gas/src/call_graph/sort.rs

```diff
@@ -0,0 +1,140 @@
+///
+/// In case the are no cycles - a Topological sorted `Vec` of `T` (the type of a `Node` value) are returned.
+/// Otherwise, an error stating there are cycles in the `CallGraph` are returned.
+///
+/// When `return_cycles = true` a specific cycles is returned. This is handy for debugging / testing purposes.
+/// However, it should not be used in production for two reasons:
+///
+/// * It's a waste of CPU cycles. An invalid Wasm program should be discarded from the mem-pool at once.
+/// * The current code for deriving a cycles is recursive (it's much easier to code a DFS search using recursions).
+///   Malicious Wasm programs might contain very big cycles which has the potential to exhaust the Stack when recursing.
+pub fn try_topological_sort<T>(
+    call_graph: &CallGraph<T>,
+    return_cycles: bool,
+) -> Result<Vec<T>, ProgramError<T>>
+where
+    T: Value,
+{
+    let mut sorted = Vec::new();
+    let mut queued = source_nodes(call_graph);
+
+    while queued.is_empty() == false {
+        let source = queued.pop().unwrap();
+
+        sorted.push(source.value());
+
+        for dest in source.outgoing() {
+            let edge = (source.value(), dest.value());
+
+            call_graph.remove_edge(edge);
+
+            if dest.is_source() {
+                queued.push(dest);
+            }
+        }
+    }
+
+    if has_edges(call_graph) {
+        let cycle = derive_cycle(call_graph, return_cycles);
+
+        Err(ProgramError::CallCycle(cycle))
+    } else {
+        Ok(sorted)
+    }
+}
+
+fn source_nodes<T>(call_graph: &CallGraph<T>) -> Vec<NodeRef<T>>
+where
+    T: Value,
+{
+    call_graph
+        .nodes()
+        .iter()
+        .filter(|node| node.is_source())
+        .cloned()
+        .collect()
+}
+
+fn has_edges<T>(call_graph: &CallGraph<T>) -> bool
+where
+    T: Value,
+{
+    call_graph.nodes().iter().any(|node| {
+        let isolated = node.is_source() && node.is_sink();
+
+        (!isolated)
+    })
+}
+
+fn derive_cycle<T>(call_graph: &CallGraph<T>, return_cycles: bool) -> Option<Vec<T>>
+where
+    T: Value,
+{
+    // The `return_cycles` feature is optional.
+    // In production we should have it turned-off.
+    if !return_cycles {
+        return None;
+    }
+
+    // Since, we know there is at least one cycle within the `CallGraph`.
+    // Each cycle's node must have at least one outgoing and one incoming edges.
+    let mut candidates = call_graph
+        .nodes()
+        .iter()
+        .filter(|node| node.has_incoming() && node.has_outgoing())
+        .cloned()
+        .collect::<Vec<_>>();
+
+    // We order so that the tests will be deterministic.
+    // So we always pick the first node out of the sorted `candidates`
+    candidates.sort_by_key(|node| node.value());
+
+    let first = candidates.first().unwrap();
+
+    let mut cycle = Vec::new();
+    let mut visited = HashSet::new();
+
+    let found = find_cycle(&first, &mut cycle, &mut visited);
```

Comment:
Perhaps re-adjust `find_cycle` signature by having a separate one for the recursion, and one for the external usage.